crates/sui-types/src/proto_value.rs

```diff
@@ -0,0 +1,1412 @@
+            // Base64 encode arbitrary bytes
+            use base64::{engine::general_purpose::STANDARD, Engine};
+
+            let bytes = BytesVisitor.visit_vector(driver)?;
+            let b64 = STANDARD.encode(bytes);
+            self.debit_string_value(&b64)?;
+            Value::from(b64)
+        } else {
+            let mut elems = vec![];
+            self.debit_value()?;
+            while let Some(elem) = driver.next_element(self)? {
+                elems.push(elem);
+            }
+
+            Value::from(elems)
+        };
+
+        self.dec_depth();
+        Ok(value)
+    }
+
+    fn visit_struct(&mut self, driver: &mut StructDriver<'_, 'b, 'l>) -> Result<Value, Error> {
+        self.inc_depth()?;
+
+        let ty = &driver.struct_layout().type_;
+
+        let value = if ty == &move_ascii_str_layout().type_ || ty == &move_utf8_str_layout().type_ {
+            // 0x1::ascii::String, 0x1::string::String
+            let s = StringVisitor.visit_struct(driver)?;
+            self.debit_string_value(s)?;
+            Value::from(s)
+        } else if ty == &UID::type_() {
+            // 0x2::object::UID
+            let uid = UidVisitor.visit_struct(driver)?.to_canonical_string(true);
+            self.debit_string_value(&uid)?;
+            Value::from(uid)
+        } else if ty == &ID::type_() {
+            // 0x2::object::ID
+            let id = IdVisitor.visit_struct(driver)?.to_canonical_string(true);
+            self.debit_string_value(&id)?;
+            Value::from(id)
+        } else if (&ty.address, ty.module.as_ref(), ty.name.as_ref()) == RESOLVED_STD_OPTION {
+            // 0x1::option::Option
+            self.debit_value()?;
+            match OptionVisitor(self).visit_struct(driver)? {
+                Some(value) => value,
+                None => Kind::NullValue(0).into(),
+            }
+        } else if Balance::is_balance(ty) {
+            // 0x2::balance::Balance
+            let balance = BalanceVisitor.visit_struct(driver)?.to_string();
+            self.debit_string_value(&balance)?;
+            Value::from(balance)
+        } else if is_url_type(ty) {
+            // 0x2::url::Url
+            let url = UrlVisitor.visit_struct(driver)?;
+            self.debit_string_value(url)?;
+            Value::from(url)
+        } else {
+            // Arbitrary structs
+            let mut map = Struct::default();
+
+            self.debit_value()?;
+            for field in &driver.struct_layout().fields {
+                self.debit_str(field.name.as_str())?;
+            }
+
+            while let Some((field, elem)) = driver.next_field(self)? {
+                map.fields.insert(field.name.as_str().to_owned(), elem);
+            }
+            Value::from(Kind::StructValue(map))
+        };
+
+        self.dec_depth();
+        Ok(value)
+    }
+
+    fn visit_variant(&mut self, driver: &mut VariantDriver<'_, 'b, 'l>) -> Result<Value, Error> {
```

Comment:
We represent enum variants a bit differently in GraphQL:

- In the `json` representation, they are represented as a singleton object whose only field's name is the variant name:

```js
{ "@variant": "Foo", "bar": "baz", "qux": "quy" } // becomes
{ "Foo": { "bar": "baz", "qux": "quy" } }
```

- In the `data` representation, we spell it out a bit more:

```js
{ "Variant": { "name": "Foo", "fields": [...] } }
```

Probably you would hew closer to the `json` format here. I don't have a particular preference between the formats, but it would be good to align. The GraphQL format was chosen to match how we tend to prefer representing enums in the TypeScript SDK.